File: af65da-Expand_Test_Coverage_for_AdjacentSum_Edge_Cases/tests/meta_test.py
#!/usr/bin/env python3
"""
Meta test for AdjacentSumTest.java test suite.

This script validates that the test suite meets the following requirements:
1. Tests for empty and single-element arrays
2. Tests for arrays containing both positive and negative integers
3. Tests for arrays containing zero values in various positions
4. Tests for integer overflow and underflow behavior
5. Tests for boundary and extreme input values

Usage: python3 meta_test.py <repository_path>
Example: python3 meta_test.py repository_before
         python3 meta_test.py repository_after
"""

import os
import re
import subprocess
import sys


def parse_test_file(test_file_path):
    """Parse the Java test file and extract test method names and their contents."""
    with open(test_file_path, 'r') as f:
        content = f.read()

    # Find all test methods with their bodies (handle nested braces)
    test_methods = {}

    # Pattern to find @Test annotations followed by method declarations
    pattern = r'@Test\s+public\s+void\s+(\w+)\s*\(\s*\)\s*\{'

    for match in re.finditer(pattern, content):
        method_name = match.group(1)
        start_pos = match.end()

        # Find matching closing brace
        brace_count = 1
        pos = start_pos
        while brace_count > 0 and pos < len(content):
            if content[pos] == '{':
                brace_count += 1
            elif content[pos] == '}':
                brace_count -= 1
            pos += 1

        method_body = content[start_pos:pos-1]
        test_methods[method_name] = method_body

    return test_methods


def check_requirement_1(tests):
    """
    Requirement 1: The test suite shall verify correct behavior of AdjacentSum 
    when given empty and single-element arrays.

    Needs BOTH:
    - Test for empty array: int[] arr = {}
    - Test for single element array: int[] arr = {n} (single element)
    """
    has_empty_array_test = False
    has_single_element_test = False

    for name, body in tests.items():
        name_lower = name.lower()

        # Check for empty array test - explicit empty array {}
        if 'empty' in name_lower:
            has_empty_array_test = True
        elif re.search(r'int\s*\[\s*\]\s*\w+\s*=\s*\{\s*\}', body):
            has_empty_array_test = True

        # Check for single element array test
        if 'single' in name_lower:
            has_single_element_test = True
        # Pattern for single element: {number} where number has no comma after
        elif re.search(r'int\s*\[\s*\]\s*\w+\s*=\s*\{\s*-?\d+\s*\}', body):
            # Make sure it's not followed by comma (truly single element)
            if not re.search(r'int\s*\[\s*\]\s*\w+\s*=\s*\{\s*-?\d+\s*,', body):
                has_single_element_test = True

    return has_empty_array_test and has_single_element_test


def check_requirement_2(tests):
    """
    Requirement 2: The test suite shall validate AdjacentSum results for arrays 
    containing both positive and negative integers.

    Needs tests that have arrays with BOTH positive AND negative numbers in same array.
    """
    has_mixed_positive_negative_test = False

    for name, body in tests.items():
        name_lower = name.lower()

        # Direct indicator in name
        if 'mixed' in name_lower and ('positive' in name_lower or 'negative' in name_lower):
            has_mixed_positive_negative_test = True
            break

        # Find arrays and check if they contain both positive and negative
        array_matches = re.findall(r'\{([^}]+)\}', body)
        for arr_content in array_matches:
            # Check for presence of negative numbers
            has_negative = bool(re.search(r'-\d+', arr_content))
            # Check for presence of positive numbers (not part of negative)
            # Remove negative numbers first, then check for remaining digits
            without_negative = re.sub(r'-\d+', '', arr_content)
            has_positive = bool(re.search(r'\d+', without_negative))

            if has_negative and has_positive:
                has_mixed_positive_negative_test = True
                break

        if has_mixed_positive_negative_test:
            break

    return has_mixed_positive_negative_test


def check_requirement_3(tests):
    """
    Requirement 3: The test suite shall confirm correct handling of arrays
    containing zero values in various positions.

    Needs explicit test for arrays containing 0 values.
    """
    has_zero_test = False

    for name, body in tests.items():
        name_lower = name.lower()

        # Check if test name indicates zero testing
        if 'zero' in name_lower:
            has_zero_test = True
            break

        # Check for arrays containing 0 (as a standalone number, not part of larger number like 10)
        # Pattern: {0, or , 0, or , 0} - zero as array element
        if re.search(r'\{\s*0\s*,', body) or re.search(r',\s*0\s*,', body) or re.search(r',\s*0\s*\}', body):
            has_zero_test = True
            break

    return has_zero_test


def check_requirement_4(tests):
    """
    Requirement 4: The test suite shall include tests that exercise integer 
    overflow and underflow behavior based on Java int arithmetic.

    Needs BOTH:
    - Overflow test: Integer.MAX_VALUE + positive number
    - Underflow test: Integer.MIN_VALUE + negative number
    """
    has_overflow_test = False
    has_underflow_test = False

    for name, body in tests.items():
        name_lower = name.lower()

        # Check for overflow test
        if 'overflow' in name_lower:
            has_overflow_test = True
        # Integer.MAX_VALUE with a positive addend causes overflow
        if 'Integer.MAX_VALUE' in body:
            # Check if paired with positive number that would cause overflow
            if re.search(r'Integer\.MAX_VALUE\s*,\s*[1-9]', body) or re.search(r'[1-9]\s*,\s*Integer\.MAX_VALUE', body):
                has_overflow_test = True

        # Check for underflow test
        if 'underflow' in name_lower:
            has_underflow_test = True
        # Integer.MIN_VALUE with a negative addend causes underflow
        if 'Integer.MIN_VALUE' in body:
            # Check if paired with negative number that would cause underflow
            if re.search(r'Integer\.MIN_VALUE\s*,\s*-[1-9]', body) or re.search(r'-[1-9]\s*,\s*Integer\.MIN_VALUE', body):
                has_underflow_test = True

    return has_overflow_test and has_underflow_test


def check_requirement_5(tests):
    """
    Requirement 5: The test suite shall ensure consistent and well-defined behavior 
    for boundary and extreme input values to prevent future regressions.

    Needs tests for boundary values:
    - Integer.MAX_VALUE boundary test
    - Integer.MIN_VALUE boundary test
    """
    has_max_value_test = False
    has_min_value_test = False

    for name, body in tests.items():
        # Check for Integer.MAX_VALUE boundary test
        if 'Integer.MAX_VALUE' in body:
            has_max_value_test = True

        # Check for Integer.MIN_VALUE boundary test
        if 'Integer.MIN_VALUE' in body:
            has_min_value_test = True

    return has_max_value_test and has_min_value_test


def compile_and_run_tests(repo_path):
    """Compile Java files and run JUnit tests."""
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    full_repo_path = os.path.join(base_dir, repo_path)

    src_main = os.path.join(full_repo_path, 'src', 'main', 'java')
    src_test = os.path.join(full_repo_path, 'src', 'test')
    lib_dir = os.path.join(base_dir, 'lib')
    build_dir = os.path.join(full_repo_path, 'build')

    # Check if lib directory exists (for Docker environment)
    if not os.path.exists(lib_dir):
        print(f"Library directory not found: {lib_dir}")
        print("Skipping compilation (JUnit jars not available)")
        return None, "Compilation skipped - running in environment without JUnit jars"

    # Create build directory
    os.makedirs(build_dir, exist_ok=True)

    # Classpath separator (: for Unix, ; for Windows)
    cp_sep = ':' if os.name != 'nt' else ';'

    junit_jar = os.path.join(lib_dir, 'junit-4.13.2.jar')
    hamcrest_jar = os.path.join(lib_dir, 'hamcrest-core-1.3.jar')

    if not os.path.exists(junit_jar) or not os.path.exists(hamcrest_jar):
        print("JUnit jars not found in lib directory")
        return None, "Compilation skipped - JUnit jars not found"

    # Compile main source
    compile_main_cmd = [
        'javac', '-d', build_dir,
        os.path.join(src_main, 'AdjacentSum.java')
    ]

    result = subprocess.run(compile_main_cmd, capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Failed to compile main source: {result.stderr}")
        return False, result.stderr

    # Compile test source
    classpath = f"{build_dir}{cp_sep}{junit_jar}{cp_sep}{hamcrest_jar}"
    compile_test_cmd = [
        'javac', '-cp', classpath, '-d', build_dir,
        os.path.join(src_test, 'AdjacentSumTest.java')
    ]

    result = subprocess.run(compile_test_cmd, capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Failed to compile test source: {result.stderr}")
        return False, result.stderr

    # Run tests
    run_classpath = f"{build_dir}{cp_sep}{junit_jar}{cp_sep}{hamcrest_jar}"
    run_cmd = [
        'java', '-cp', run_classpath,
        'org.junit.runner.JUnitCore', 'AdjacentSumTest'
    ]

    result = subprocess.run(run_cmd, capture_output=True, text=True)

    # Parse test results
    output = result.stdout + result.stderr

    # Check if all tests passed
    all_passed = 'OK' in output and 'FAILURES' not in output

    return all_passed, output


def run_meta_tests(repo_path):
    """Run all meta tests on the given repository."""
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    test_file_path = os.path.join(base_dir, repo_path, 'src', 'test', 'AdjacentSumTest.java')

    if not os.path.exists(test_file_path):
        print(f"Error: Test file not found at {test_file_path}")
        return False

    print(f"Analyzing test file: {test_file_path}")
    print("=" * 60)

    tests = parse_test_file(test_file_path)

    print(f"Found {len(tests)} test methods:")
    for name in tests.keys():
        print(f"  - {name}")
    print("=" * 60)

    results = []

    # Check each requirement
    req1_passed = check_requirement_1(tests)
    results.append(('REQ-1: Empty and single-element arrays', req1_passed))

    req2_passed = check_requirement_2(tests)
    results.append(('REQ-2: Positive and negative integers', req2_passed))

    req3_passed = check_requirement_3(tests)
    results.append(('REQ-3: Arrays containing zero values', req3_passed))

    req4_passed = check_requirement_4(tests)
    results.append(('REQ-4: Integer overflow and underflow', req4_passed))

    req5_passed = check_requirement_5(tests)
    results.append(('REQ-5: Boundary and extreme values', req5_passed))

    # Print results
    print("\nRequirement Coverage Results:")
    print("-" * 60)

    passed_count = 0
    for req_name, passed in results:
        status = "PASS" if passed else "FAIL"
        print(f"  {req_name}: {status}")
        if passed:
            passed_count += 1

    print("-" * 60)
    print(f"Total: {passed_count}/{len(results)} requirements covered")
    print("=" * 60)

    # Try to compile and run the actual tests
    print("\nCompiling and running JUnit tests...")
    print("-" * 60)

    tests_passed, test_output = compile_and_run_tests(repo_path)

    if test_output:
        print("Test execution output:")
        if isinstance(test_output, list):
            for line in test_output:
                print(f"  {line}")
        else:
            for line in test_output.split('\n'):
                print(f"  {line}")

    print("-" * 60)

    if tests_passed is None:
        print("JUnit tests: SKIPPED (environment not configured)")
    elif tests_passed:
        print("JUnit tests: ALL PASSED")
    else:
        print("JUnit tests: SOME FAILED (or compilation error)")

    print("=" * 60)

    # Final verdict
    all_requirements_met = all(passed for _, passed in results)

    print("\nFinal Verdict:")
    if all_requirements_met:
        print("  SUCCESS: All requirements are covered by the test suite.")
        return True
    else:
        print("  FAILURE: Not all requirements are covered by the test suite.")
        missing = [name for name, passed in results if not passed]
        print(f"  Missing coverage for: {', '.join(missing)}")
        return False


def main():
    if len(sys.argv) != 2:
        print("Usage: python3 meta_test.py <repository_path>")
        print("Example: python3 meta_test.py repository_before")
        print("         python3 meta_test.py repository_after")
        sys.exit(1)

    repo_path = sys.argv[1]

    # Validate repository path
    if repo_path not in ['repository_before', 'repository_after']:
        print(f"Warning: Unexpected repository path '{repo_path}'")
        print("Expected 'repository_before' or 'repository_after'")

    print(f"\nRunning meta tests for: {repo_path}")
    print("=" * 60)

    success = run_meta_tests(repo_path)

    # Exit with appropriate code
    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

File: af65da-Expand_Test_Coverage_for_AdjacentSum_Edge_Cases/tests/MetaTest.java
import org.junit.Test;
import static org.junit.Assert.*;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.ArrayList;
import java.util.List;

public class MetaTest {
    
    // ========================================================================
    // SECTION 1: CLASS CONSTANTS
    // ========================================================================
    
    /** 
     * Minimum number of test methods required for full coverage.
     * The repository_after should have at least 10 tests to cover all scenarios.
     */
    private static final int MINIMUM_REQUIRED_TESTS = 10;
    
    /** 
     * Reference to the AdjacentSumTest class being analyzed.
     * This is loaded dynamically to allow testing different versions.
     */
    private static Class<?> testClass;
    
    /** Set of all test method names (lowercase for case-insensitive matching) */
    private static Set<String> testMethodNames;
    
    /** List to track which requirements passed/failed */
    private static List<String[]> requirementResults;
    
    // ========================================================================
    // SECTION 2: STATIC INITIALIZATION
    // ========================================================================
    
    static {
        try {
            // Step 1: Load the test class dynamically
            // This allows testing different versions of AdjacentSumTest
            testClass = Class.forName("AdjacentSumTest");
            
            // Step 2: Get all declared methods using reflection
            Method[] methods = testClass.getDeclaredMethods();
            
            // Step 3: Filter and collect @Test annotated method names
            testMethodNames = new HashSet<>();
            for (Method method : methods) {
                // Check if method has @Test annotation
                if (method.isAnnotationPresent(Test.class)) {
                    // Store lowercase name for case-insensitive matching
                    testMethodNames.add(method.getName().toLowerCase());
                }
            }
            
            // Initialize results tracking
            requirementResults = new ArrayList<>();
            
        } catch (ClassNotFoundException e) {
            // If AdjacentSumTest is not on classpath, tests will fail gracefully
            testClass = null;
            testMethodNames = new HashSet<>();
            requirementResults = new ArrayList<>();
            System.err.println("WARNING: AdjacentSumTest class not found on classpath");
        }
    }
    
    // ========================================================================
    // SECTION 3: MAIN ENTRY POINT (Standalone Execution)
    // ========================================================================
    
    public static void main(String[] args) {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("META-TEST: AdjacentSumTest Coverage Analysis (Reflection-Based)");
        System.out.println("=".repeat(70));
        
        // Check if test class was loaded successfully
        if (testClass == null) {
            System.out.println("\nERROR: AdjacentSumTest class not found!");
            System.out.println("Make sure AdjacentSumTest.class is on the classpath.");
            System.exit(1);
        }
        
        // Print discovered test methods
        System.out.println("\nAnalyzing class: " + testClass.getName());
        System.out.println("Found " + testMethodNames.size() + " @Test methods:");
        System.out.println("-".repeat(70));
        
        // Get actual method names (not lowercase) for display
        for (Method method : testClass.getDeclaredMethods()) {
            if (method.isAnnotationPresent(Test.class)) {
                System.out.println("  - " + method.getName() + "()");
            }
        }
        
        System.out.println("-".repeat(70));
        
        // Create instance and run checks
        MetaTest metaTest = new MetaTest();
        
        // Clear previous results
        requirementResults.clear();
        
        // Run all requirement checks
        System.out.println("\nRequirement Coverage Analysis:");
        System.out.println("-".repeat(70));
        
        boolean req1 = metaTest.checkRequirement1_EmptyAndSingleElement();
        boolean req2 = metaTest.checkRequirement2_MixedPositiveNegative();
        boolean req3 = metaTest.checkRequirement3_ZeroValues();
        boolean req4 = metaTest.checkRequirement4_OverflowUnderflow();
        boolean req5 = metaTest.checkRequirement5_BoundaryValues();
        boolean reqCount = metaTest.checkMinimumTestCount();
        
        // Print results summary
        System.out.println("\n" + "-".repeat(70));
        System.out.println("RESULTS SUMMARY:");
        System.out.println("-".repeat(70));
        
        for (String[] result : requirementResults) {
            String status = result[1].equals("PASS") ? "PASS" : "FAIL";
            System.out.println("  " + status + " - " + result[0]);
        }
        
        // Calculate overall result
        int passedCount = 0;
        for (String[] result : requirementResults) {
            if (result[1].equals("PASS")) passedCount++;
        }
        
        System.out.println("-".repeat(70));
        System.out.println("Total: " + passedCount + "/" + requirementResults.size() + " checks passed");
        System.out.println("=".repeat(70));
        
        // Final verdict (requirements 1-5 must all pass)
        boolean allRequirementsPassed = req1 && req2 && req3 && req4 && req5;
        
        System.out.println("\nFINAL VERDICT:");
        if (allRequirementsPassed) {
            System.out.println("  SUCCESS: All requirements are covered by the test suite.");
            System.exit(0);
        } else {
            System.out.println("  FAILURE: Not all requirements are covered by the test suite.");
            
            // List missing requirements
            System.out.print("  Missing: ");
            List<String> missing = new ArrayList<>();
            if (!req1) missing.add("REQ-1 (Empty/Single)");
            if (!req2) missing.add("REQ-2 (Mixed +/-)");
            if (!req3) missing.add("REQ-3 (Zeros)");
            if (!req4) missing.add("REQ-4 (Overflow/Underflow)");
            if (!req5) missing.add("REQ-5 (Boundaries)");
            System.out.println(String.join(", ", missing));
            
            System.exit(1);
        }
    }
    
    // ========================================================================
    // SECTION 4: REQUIREMENT 1 - Empty and Single-Element Arrays
    // ========================================================================
    
    private boolean checkRequirement1_EmptyAndSingleElement() {
        // Check for "empty" in any test method name
        boolean hasEmptyArrayTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("empty"));
        
        // Check for "single" in any test method name
        boolean hasSingleElementTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("single"));
        
        // Record result
        boolean passed = hasEmptyArrayTest && hasSingleElementTest;
        requirementResults.add(new String[]{
            "REQ-1: Empty and single-element arrays", 
            passed ? "PASS" : "FAIL"
        });
        
        // Print detailed status
        System.out.println("  REQ-1: Empty and single-element arrays");
        System.out.println("         - Empty array test: " + (hasEmptyArrayTest ? "Found" : "MISSING"));
        System.out.println("         - Single element test: " + (hasSingleElementTest ? "Found" : "MISSING"));
        System.out.println("         Result: " + (passed ? "PASS" : "FAIL"));
        
        return passed;
    }
    
    /**
     * JUnit test method for Requirement 1.
     * This is called when running via JUnit runner.
     */
    @Test
    public void testRequirement1_EmptyAndSingleElementArrays() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        boolean hasEmptyArrayTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("empty"));
        boolean hasSingleElementTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("single"));
        
        assertTrue("Missing test for empty array (Requirement 1)", hasEmptyArrayTest);
        assertTrue("Missing test for single element array (Requirement 1)", hasSingleElementTest);
    }
    
    // ========================================================================
    // SECTION 5: REQUIREMENT 2 - Mixed Positive and Negative Integers
    // ========================================================================
    
    private boolean checkRequirement2_MixedPositiveNegative() {
        boolean hasMixedSignTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("mixed") || 
                             (name.contains("negative") && name.contains("positive")));
        
        // Also check for explicit mixed naming
        boolean hasExplicitMixedTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("mixedpositive") || 
                             name.contains("mixednegative") ||
                             name.contains("positivenegative") ||
                             name.contains("negativepositive"));
        
        boolean passed = hasMixedSignTest || hasExplicitMixedTest;
        requirementResults.add(new String[]{
            "REQ-2: Mixed positive and negative integers", 
            passed ? "PASS" : "FAIL"
        });
        
        System.out.println("  REQ-2: Mixed positive and negative integers");
        System.out.println("         - Mixed sign test: " + (passed ? "Found" : "MISSING"));
        System.out.println("         Result: " + (passed ? "PASS" : "FAIL"));
        
        return passed;
    }
    
    /**
     * JUnit test method for Requirement 2.
     */
    @Test
    public void testRequirement2_MixedPositiveNegativeIntegers() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        boolean hasMixedSignTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("mixed") || 
                             (name.contains("negative") && name.contains("positive")));
        
        assertTrue("Missing test for mixed positive/negative integers (Requirement 2)", 
                   hasMixedSignTest);
    }
    
    // ========================================================================
    // SECTION 6: REQUIREMENT 3 - Zero Values
    // ========================================================================
    
    private boolean checkRequirement3_ZeroValues() {
        boolean hasZeroTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("zero"));
        
        requirementResults.add(new String[]{
            "REQ-3: Arrays containing zero values", 
            hasZeroTest ? "PASS" : "FAIL"
        });
        
        System.out.println("  REQ-3: Arrays containing zero values");
        System.out.println("         - Zero values test: " + (hasZeroTest ? "Found" : "MISSING"));
        System.out.println("         Result: " + (hasZeroTest ? "PASS" : "FAIL"));
        
        return hasZeroTest;
    }
    
    /**
     * JUnit test method for Requirement 3.
     */
    @Test
    public void testRequirement3_ZeroValues() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        boolean hasZeroTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("zero"));
        
        assertTrue("Missing test for zero values (Requirement 3)", hasZeroTest);
    }
    
    // ========================================================================
    // SECTION 7: REQUIREMENT 4 - Integer Overflow and Underflow
    // ========================================================================

    private boolean checkRequirement4_OverflowUnderflow() {
        boolean hasOverflowTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("overflow"));
        
        boolean hasUnderflowTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("underflow"));
        
        boolean passed = hasOverflowTest && hasUnderflowTest;
        requirementResults.add(new String[]{
            "REQ-4: Integer overflow and underflow", 
            passed ? "PASS" : "FAIL"
        });
        
        System.out.println("  REQ-4: Integer overflow and underflow");
        System.out.println("         - Overflow test: " + (hasOverflowTest ? "Found" : "MISSING"));
        System.out.println("         - Underflow test: " + (hasUnderflowTest ? "Found" : "MISSING"));
        System.out.println("         Result: " + (passed ? "PASS" : "FAIL"));
        
        return passed;
    }
    
    /**
     * JUnit test method for Requirement 4.
     */
    @Test
    public void testRequirement4_IntegerOverflowUnderflow() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        boolean hasOverflowTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("overflow"));
        boolean hasUnderflowTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("underflow"));
        
        assertTrue("Missing test for integer overflow (Requirement 4)", hasOverflowTest);
        assertTrue("Missing test for integer underflow (Requirement 4)", hasUnderflowTest);
    }
    
    // ========================================================================
    // SECTION 8: REQUIREMENT 5 - Boundary and Extreme Values
    // ========================================================================
    
    private boolean checkRequirement5_BoundaryValues() {
        boolean hasMaxValueTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("max"));
        
        boolean hasMinValueTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("min"));
        
        boolean passed = hasMaxValueTest && hasMinValueTest;
        requirementResults.add(new String[]{
            "REQ-5: Boundary and extreme values", 
            passed ? "PASS" : "FAIL"
        });
        
        System.out.println("  REQ-5: Boundary and extreme values");
        System.out.println("         - MAX_VALUE test: " + (hasMaxValueTest ? "Found" : "MISSING"));
        System.out.println("         - MIN_VALUE test: " + (hasMinValueTest ? "Found" : "MISSING"));
        System.out.println("         Result: " + (passed ? "PASS" : "FAIL"));
        
        return passed;
    }
    
    /**
     * JUnit test method for Requirement 5.
     */
    @Test
    public void testRequirement5_BoundaryValues() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        boolean hasMaxValueTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("max"));
        boolean hasMinValueTest = testMethodNames.stream()
            .anyMatch(name -> name.contains("min"));
        
        assertTrue("Missing test for MAX_VALUE boundary (Requirement 5)", hasMaxValueTest);
        assertTrue("Missing test for MIN_VALUE boundary (Requirement 5)", hasMinValueTest);
    }
    
    // ========================================================================
    // SECTION 9: MINIMUM TEST COUNT CHECK
    // ========================================================================
    
    private boolean checkMinimumTestCount() {
        int testCount = testMethodNames.size();
        boolean passed = testCount >= MINIMUM_REQUIRED_TESTS;
        
        requirementResults.add(new String[]{
            "Minimum test count (>=" + MINIMUM_REQUIRED_TESTS + ")", 
            passed ? "PASS" : "FAIL"
        });
        
        System.out.println("  Minimum test count check:");
        System.out.println("         - Found: " + testCount + " tests");
        System.out.println("         - Required: >= " + MINIMUM_REQUIRED_TESTS + " tests");
        System.out.println("         Result: " + (passed ? "PASS" : "FAIL"));
        
        return passed;
    }
    
    /**
     * JUnit test method for minimum test count.
     */
    @Test
    public void testMinimumNumberOfTests() {
        assertNotNull("AdjacentSumTest class not found", testClass);
        
        long testCount = testMethodNames.size();
        
        assertTrue("Test suite should have at least " + MINIMUM_REQUIRED_TESTS + 
                   " tests for full coverage, but found only " + testCount, 
                   testCount >= MINIMUM_REQUIRED_TESTS);
    }
    
    // ========================================================================
    // SECTION 10: BEHAVIOR-BASED META-TESTS (Optional)
    // ========================================================================

    @Test
    public void metaTest_EmptyArrayTestExists() {
        try {
            // Try to find and invoke testEmptyArray
            Object testInstance = testClass.getDeclaredConstructor().newInstance();
            Method emptyArrayTest = testClass.getMethod("testEmptyArray");
            emptyArrayTest.invoke(testInstance);
            // If we get here, the test exists and runs
        } catch (NoSuchMethodException e) {
            fail("Test suite missing testEmptyArray() - Requirement 1 not covered");
        } catch (Exception e) {
            // Other exceptions are OK - the method exists but might have assertions
            // that fail (which is expected behavior for a test)
        }
    }
    
    @Test
    public void metaTest_OverflowTestExists() {
        try {
            Object testInstance = testClass.getDeclaredConstructor().newInstance();
            Method overflowTest = testClass.getMethod("testIntegerOverflow");
            overflowTest.invoke(testInstance);
        } catch (NoSuchMethodException e) {
            fail("Test suite missing testIntegerOverflow() - Requirement 4 not covered");
        } catch (Exception e) {
            // Method exists
        }
    }
    
    @Test
    public void metaTest_UnderflowTestExists() {
        try {
            Object testInstance = testClass.getDeclaredConstructor().newInstance();
            Method underflowTest = testClass.getMethod("testIntegerUnderflow");
            underflowTest.invoke(testInstance);
        } catch (NoSuchMethodException e) {
            fail("Test suite missing testIntegerUnderflow() - Requirement 4 not covered");
        } catch (Exception e) {
            // Method exists
        }
    }
    
    @Test
    public void metaTest_ZeroValuesTestExists() {
        try {
            Object testInstance = testClass.getDeclaredConstructor().newInstance();
            Method zeroTest = testClass.getMethod("testArrayWithZeros");
            zeroTest.invoke(testInstance);
        } catch (NoSuchMethodException e) {
            fail("Test suite missing testArrayWithZeros() - Requirement 3 not covered");
        } catch (Exception e) {
            // Method exists
        }
    }
    
    @Test
    public void metaTest_MixedPositiveNegativeTestExists() {
        try {
            Object testInstance = testClass.getDeclaredConstructor().newInstance();
            Method mixedTest = testClass.getMethod("testMixedPositiveAndNegative");
            mixedTest.invoke(testInstance);
        } catch (NoSuchMethodException e) {
            fail("Test suite missing testMixedPositiveAndNegative() - Requirement 2 not covered");
        } catch (Exception e) {
            // Method exists
        }
    }
}
